### DISCUSSION TITLE
Troubleshooting OpenLiteSpeed with Django and Daphne: Overcoming 503 Errors

### DISCUSSION THREAD

**Discussion by tawfeeqzaghlool** [community/community#137920] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Problem
I recently moved to OpenLiteSpeed Django and have my repository ready to be cloned. I have implemented Daphne to work smoothly with ASGI. However, I am encountering a 503 Service Unavailable error. I am unsure if OpenLiteSpeed can be used without PHP, as it seems to be waiting for PHP to run or is built to work within a PHP environment.

### What I Have Tried
1. **Setup**: 
   - Cloned my repository and set up everything needed for OpenLiteSpeed.
   - Created a Daphne service file and configured it to create a socket file for OpenLiteSpeed to communicate with.
   - Ensured all permissions are set correctly, and verified that the OpenLiteSpeed user "nobody" can access the socket file.

2. **Logs**:
   - Checked all logs (access, error, debug) but found no useful information.
   - The only consistent issue is the 503 Service Unavailable error.

3. **Configuration**:
   - Removed the Example app block code from the `hosts_config.conf` file.
   - Tested the syntax with the `-t` command, and no errors were found.

### Question
Can I use OpenLiteSpeed without PHP for my Django app? If not, what could be the best alternative to OpenLiteSpeed that works with ASGI apps and Daphne.

**Comment by shruti-devganiy-3024** [comment]
Yes, OpenLiteSpeed can run Django/ASGI applications without PHP. PHP is not required for OpenLiteSpeed itself.

The 503 usually means OpenLiteSpeed cannot communicate with the upstream Daphne process correctly, not that PHP is missing.

Since you already confirmed:

* Daphne service is running
* socket permissions look correct
* config syntax passes validation

then the issue is most likely one of these:

1. Socket path mismatch between Daphne and OpenLiteSpeed
2. OpenLiteSpeed user cannot actually access parent directories
3. Daphne is binding differently than expected
4. OpenLiteSpeed external app configuration is incorrect

One thing that catches people often is that even if the socket file itself is readable, OpenLiteSpeed (`nobody`) also needs execute permission on every parent directory leading to the socket.

For example:

```bash id="jlr6p2"
chmod o+x /path/to/socket-parent-dir
```

Also verify Daphne is actually listening on the expected socket:

```bash id="fdyjlwm"
sudo ss -lx | grep daphne
```

or test temporarily using TCP instead of Unix sockets just to isolate whether the problem is socket-related:

```bash id="8d3qep"
daphne -b 127.0.0.1 -p 8001 project.asgi:application
```

and then point OpenLiteSpeed to:

```text id="8um2l7"
127.0.0.1:8001
```

If that works, the problem is almost certainly the socket configuration/permissions.

I have also seen OpenLiteSpeed return very generic 503 responses when the external app type is configured incorrectly (WSGI vs Proxy vs LSAPI), so I would double-check that the app is configured as a proxy to Daphne rather than anything PHP-related.

Honestly, if your goal is mainly stable ASGI deployment, many people end up using:

* Nginx + Daphne
* Nginx + Uvicorn
* Caddy + Uvicorn

because the ecosystem/documentation around ASGI is much more mature there compared to OpenLiteSpeed.

**Comment by sairamkollur** [comment]
Yes, OpenLiteSpeed can absolutely run Django applications without PHP. PHP is not required for OpenLiteSpeed itself — it is just commonly bundled because many users host WordPress or PHP-based apps.

Your 503 error is most likely caused by OpenLiteSpeed failing to communicate with Daphne through the Unix socket or upstream configuration, not because PHP is missing.

A few things to verify:

1. Confirm Daphne is actually running:

```bash
systemctl status daphne
```

2. Test the socket manually:

```bash
curl --unix-socket /path/to/daphne.sock http://localhost/
```

3. Check that the socket permissions allow the OpenLiteSpeed user (`nobody`) to access it:

```bash
ls -la /path/to/daphne.sock
```

4. In OpenLiteSpeed, ensure you configured:

* External App → Web Server → Proxy/LSAPI correctly
* Address matches the Daphne socket path or TCP port
* Context/Proxy rules point requests to Daphne

5. If using a Unix socket becomes problematic, temporarily switch Daphne to a TCP port:

```bash
daphne -b 127.0.0.1 -p 8001 myproject.asgi:application
```

Then proxy OpenLiteSpeed to:

```txt
127.0.0.1:8001
```

This is often easier to debug than sockets.

Also note:

* OpenLiteSpeed is more optimized/documented for PHP workflows.
* ASGI setups with Daphne/Uvicorn are generally smoother with Nginx or Caddy as reverse proxies.

Recommended production stack for Django ASGI:

```txt
Nginx → Daphne/Uvicorn → Django
```

or

```txt
Caddy → Uvicorn → Django
```

These have:

* better ASGI ecosystem support
* simpler reverse proxy configuration
* more community examples
* easier WebSocket handling

If you continue getting 503 with no meaningful logs, the issue is usually:

* upstream mismatch
* socket permission issue
* wrong proxy context
* Daphne not binding correctly
* SELinux/AppArmor restrictions
* OpenLiteSpeed trying to proxy before Daphne is ready

A quick isolation test:
Run Daphne directly on a TCP port and access it without OpenLiteSpeed first. If that works, the issue is entirely in the OpenLiteSpeed proxy configuration layer.
